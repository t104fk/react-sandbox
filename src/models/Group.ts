import { flow, Instance, types } from "mobx-state-tree";
import { TWishListItem, WishList } from "./WishList";

export const User = types.model({
  id: types.string,
  name: types.string,
  // unionも行ける
  // gender: types.union(types.literal('m'), types.literal('f')),
  gender: types.enumeration('gender', ['m', 'f']),
  wishList: types.optional(WishList, {}),
})
.actions(self => ({
  changeName(newName: string) {
    self.name = newName;
  },
  getSuggestions: flow(function*() {
    // Unhandled Rejection (Error): [mobx-state-tree] Cannot modify 'AnonymousModel[]@/users/a342/wishList/items',
    // the object is protected and can only be modified by using an action.
    // async/awaitだと中でselfをいじれないっぽい というより1行しか実行されない？
    const res = yield fetch(`http://localhost:3001/suggestions_${self.gender}`)
    // json()はanyを返すので、assertionしておいたほうがよい
    const suggestions: TWishListItem[] = yield res.json();
    // flowだと参照できる
    self.wishList.items.push(...suggestions);
    // Property 'addSuggestions' does not exist on type
    // self.addSuggestions(...(yield res.json()))
  }),
  // addSuggestions(suggestions: TWishListItem[]) {
  //   self.wishList.items.push(...suggestions)
  // }
}));
export type TUser = Instance<typeof User>;

export const Group = types.model({
  users: types.map(User)
})
export type TGroup = Instance<typeof Group>;
