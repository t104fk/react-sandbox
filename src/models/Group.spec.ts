import { getSnapshot, onSnapshot } from "mobx-state-tree";
import { User } from "./Group"

it('can create User', () => {
  const user = User.create({
    id: 'a342',
    name: 'Homer',
    gender: 'm',
  });
  const states = [];
  onSnapshot(user, snapshot => {
    states.push(snapshot);
  });

  expect(states.length).toBe(0);
  expect(user.name).toBe('Homer');
  user.changeName('Home alone');
  expect(states.length).toBe(1);
  expect(user.name).toBe('Home alone');
  expect(getSnapshot(user)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
})
