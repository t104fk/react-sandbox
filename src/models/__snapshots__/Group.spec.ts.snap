// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create User 1`] = `
Object {
  "gender": "m",
  "id": "a342",
  "name": "Home alone",
  "wishList": Object {
    "items": Array [],
  },
}
`;

exports[`can create User 2`] = `
Array [
  Object {
    "gender": "m",
    "id": "a342",
    "name": "Home alone",
    "wishList": Object {
      "items": Array [],
    },
  },
]
`;
